### DISCUSSION TITLE
Using the StructureDefinition's url:below modifier when searching

### DISCUSSION THREAD

**Discussion by ajerrison-manchester** [FirelyTeam/firely-net-sdk#1649] [Q&A] [answered]
I'm a bit of a newcomer to the Firely SDK so apologies if this has an obvious answer or not the right place to ask!

I'd like to use the SDK to search for StructureDefinitions in the same way (or similar) to the code snippet below:
```cs 
SearchParams q = new SearchParams("url:below", "https://example.com");
q.Add("kind", "resource");
q.Count = 20;
Task<Bundle> bundleTask = _conformanceClient.SearchAsync<StructureDefinition>(q);
bundle = await bundleTask;
```
I think I am not forming the url:below parameter correctly for the SDK though, as the bundle contains StructureDefinitions that are not filtered by that parameter, and more importantly it also contains an OperationOutcome error that states: _'Overall result: FAILURE (1 errors and 0 warnings)\r\n\r\n[ERROR] Parameter is not well formatted, expected 'parametername[:modifier]=value'_.

The parameter is stripped from the NextLink, SelfLink and LastLink properties in the bundle too.

Could anyone tell me how I'd need to form that parameter? Is Search() even the correct method to use? 

So far I've tried URL encoding the URL value and using the string array method of specifying the search criteria.

I don't think this is a bug, as it happens in the Java SDK too, more likely user error! A REST call using the browser with those parameters is OK and returns what I expect. I am using Nuget package HL7.Fhir.STU3 version 2.0.3.

**Comment by marcovisserFurore** [comment, marked as answer]
hi @ajerrison-manchester , thanks for using the Firely .NET SDK.

What you are trying to do should work. I've created the following unit test and that succeeds:

```c#
        [TestMethod, TestCategory("IntegrationTest"), TestCategory("FhirClient")]
        public async System.Threading.Tasks.Task Discussion1649Async()
        {
            using var client = new FhirClient("https://server.fire.ly/Administration/R3", settings: FhirClientSettings.CreateDefault());
            SearchParams q = new SearchParams("url:below", "http://hl7.org/fhir/us/core/StructureDefinition/us-core");
            q.Add("kind", "resource");
            q.Count = 20;
            var bundle = await client.SearchAsync<StructureDefinition>(q);

            Assert.IsNotNull(bundle);
            Assert.IsTrue(bundle.Entry.Count > 0);
        }
```

I use here our own Firely Fhir server: https://server.fire.ly and I use STU3, so that is why the `R3` is in the url.

Which Fhir server do you use?

**Reply by ajerrison-manchester** [reply]
Thanks for checking! I use Vonk 3.9 at present.

**Reply by ajerrison-manchester** [reply]
Hmm, using the same endpoint and parameters as you I still get the 'OperationOutcome' error, and changing the pagesize to 50 I still also get resources that are not part of 'us-core'. So there might be something else going on with our client config.
 
Your unit test would pass for me too - my bundle contains entry records but the last one in the first page is the error, and the link properties have incorrect URLs.

**Reply by marcovisserFurore** [reply]
I've found the problem: indeed I also got the `OperationOutcome` error. I overlooked that. 

When the `FhirClient` executes the request, it encodes the values. So the request looks like this:

```
https://server.fire.ly/Administration/R3/administration/StructureDefinition?_count=20&url%3Abelow=http%3A%2F%2Fhl7.org%2Ffhir%2Fus%2Fcore%2FStructureDefinition%2Fus-core&kind=resource
``` 

The Firely Server (Vonk) cannot handle this. I will notify the Firely Server Team about this

**Reply by ajerrison-manchester** [reply]
Thanks a lot, I hadn't really thought about whether the server itself was the problem! I think I can work round it while a fix is made.

**Reply by marcovisserFurore** [reply]
The Firely Team confirmed that this a bug, and put this issue on their backlog.
